File: Programming and Projection Methods/Lab2/Tema2/src/main/java/service/Service.java
package service;

import domain.Booking;
import domain.Client;
import domain.Ride;
import domain.User;
import repository.*;
import utils.*;

import java.util.ArrayList;
import java.util.List;

public class Service implements Observable<AppEvent> {
    private UserRepositoryInterface user_repo;
    private RideRepositoryInterface ride_repo;
    private ClientRepositoryInterface client_repo;
    private BookingRepositoryInterface booking_repo;
    private List<Observer<AppEvent>> observers = new ArrayList<>();

    public Service(UserRepository user_repo, RideRepository ride_repo, ClientRepository client_repo, BookingRepository booking_repo) {
        this.user_repo = user_repo;
        this.ride_repo = ride_repo;
        this.client_repo = client_repo;
        this.booking_repo = booking_repo;
    }

    public User findUser(String username, String password) {
        return user_repo.findOne(username, password);
    }

    public Iterable<RBooking> findAllRBookings(String dest, String date, String hour) {
        return booking_repo.findBookingsByDestDateHour(dest, date, hour);
    }

    public String bookPlaces(Ride ride, Client c, int nrplaces) {
        Ride old = ride;
        client_repo.save(c);
        String places = "";
        int x = nrplaces;
        String p = ride.getPlaces();
        for (int i = 1; i <= 18; i++) {
            if (ride.getPlaces().charAt(i) == '0') {
                if (x == 1)
                    places += i;
                else
                    places += i + ",";
                p = p.substring(0, i) + '1' + p.substring(i + 1);
                x--;
                if (x == 0)
                    break;
            }
        }
        ride.setPlaces(p);
        ride_repo.update(ride);
        Client cl = client_repo.findLastAdded();
        Booking booking = new Booking(cl, ride, nrplaces, places);
        booking_repo.save(booking);
        notifyObservers(new AppEvent(Action.UPDATE, ride, old));
        //notifyObservers(new AppEvent(Action.ADD,booking));
        return places;
    }

    public Iterable<Ride> findAllRides() {
        return ride_repo.findAll();
    }

    public void addObserver(Observer<AppEvent> o) {
        observers.add(o);
    }

    public void removeObserver(Observer<AppEvent> o) {
        observers.remove(o);
    }

    public void notifyObservers(AppEvent e) {
        observers.stream().forEach(x -> x.update(e));
    }
}

File: Programming and Projection Methods/Lab5C#/Lab5/Lab5/service/Service.cs
using Lab5.domain;
using Lab5.repository;
using Lab5.utils;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Lab5.service
{
    public class Service:Observable
    {
        private IUserRepository user_repo;
        private IRideRepository ride_repo;
        private IClientRepository client_repo;
        private IBookingRepository booking_repo;
        private IList<Observer> observers = new  List<Observer>();

        public Service(IUserRepository user_repo, IRideRepository ride_repo, IClientRepository client_repo, IBookingRepository booking_repo)
        {
            this.user_repo = user_repo;
            this.ride_repo = ride_repo;
            this.client_repo = client_repo;
            this.booking_repo = booking_repo;
        }

        public User findUser(String username,String password)
        {
            return user_repo.FindOne(username, password);
        }

        public IEnumerable<RBooking> findAllRBookings(String dest,String date,String hour)
        {
            return booking_repo.findBookingsByDestDateHour(dest, date, hour);
        }

        public String bookPlaces(Ride ride,Client c,int nrplaces)
        {
            ride = ride_repo.findOneby_Destination_Date_Hour(ride.Destination, ride.Date, ride.Hour);
            client_repo.Save(c);
            String places = "";
            int x = nrplaces;
            String p = ride.Places;
            for (int i = 1; i <= 18; i++)
            {
                if (ride.Places[i].Equals('0'))
                {
                    if (x == 1)
                        places += i;
                    else
                        places += i + ",";
                    p = p.Substring(0, i) + '1' + p.Substring(i + 1);
                    x--;
                    if (x == 0)
                        break;
                }
            }
            ride.Places = p;
            ride_repo.Update(ride);
            Client cl = client_repo.FindLastAdded();
            Booking booking = new Booking(new KeyValuePair<Ride,Client>(ride,cl), nrplaces, places);
            booking_repo.Save(booking);
            NotifyObservers();
            return places;
        }

        public IEnumerable<Ride> findAllRides()
        {
            return ride_repo.FindAll();
        }

        public override void AddObserver(Observer o)
        {
            observers.Add(o);
        }

        public override void NotifyObservers()
        {
            observers.ToList().ForEach(x => x.Update());
        }
    }
}
